dipy/testing/decorators.py

```diff
@@ -103,3 +105,88 @@ def _set_random_number_generator_wrapper(pytestconfig, *args, **kwargs):
 
     return _set_random_number_generator
+
+
+def warn_on_positional_args(deprecates_from_version="1.10.0"):
+    """
+    Decorator to enforce keyword-only arguments.
+
+    This decorator issues a warning if any keyword-only arguments are
+    passed positionally instead of as keyword arguments. The warning
+    indicates that such usage will be deprecated from the specified version.
+
+    Args:
+        deprecates_from_version (str): The version from which the positional
+                                       usage of keyword-only arguments will
+                                       be deprecated. Default is "1.10.0".
+
+    Returns:
+        function: The decorated function with added warnings for
+                  positional usage of keyword-only arguments.
+
+    Example:
+        @warn_on_positional_args(deprecates_from_version="2.0")
+        def add(*, a=3, b=2, c=6):
+            return a + b + c
+
+        print(add(1, 2, 3))  # This will issue a warning and reorganize arguments.
+    """
+
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            sig = inspect.signature(func)
+            param_names = list(sig.parameters.keys())
+
+            # Determine the keyword-only arguments
+            kwonly_args = [
+                name
+                for name, param in sig.parameters.items()
+                if param.kind == param.KEYWORD_ONLY
+            ]
+
+            try:
+                bound_args = sig.bind(*args, **kwargs)
+                bound_args.apply_defaults()
+            except TypeError as e:
+                if "too many positional arguments" in str(e):
+                    num_positional_args = len(args)
+                    num_expected_positional = len(param_names) - len(kwonly_args)
+                    if num_positional_args > num_expected_positional:
+                        # Move excess positional arguments to keyword arguments
+                        excess_args_start_index = num_expected_positional
+                        excess_args = args[excess_args_start_index:]
+
+                        # Collect all keyword-only arguments being called positionally
+                        positionally_passed_kwonly_args = []
+                        for i, kwarg in enumerate(kwonly_args):
+                            if i < len(excess_args):
+                                kwargs[kwarg] = excess_args[i]
+                                positionally_passed_kwonly_args.append(kwarg)
+
+                        if positionally_passed_kwonly_args:
+                            warning_msg = (
+                                f"Warning: Function '{func.__name__}' "
+                                f"keyword-only arguments "
+                                f"{positionally_passed_kwonly_args} are being called "
+                                f"positionally. This will be deprecated from version "
+                                f"'{deprecates_from_version}'."
```

Comment:
f"Pass {positionally_passed_kwonly_args} as keyword args. From version "
                    f"{deprecates_until_version} passing these as positional arguments "
                    "will result in an error"